File: phase5/scripts/p5-model-probe.py
#!/usr/bin/env python3
"""p5-model-probe.py - decide, by measurement, whether a live agent turn can run.

Why this exists: the Phase 5 re-runs of the Phase 3 drivers (G7/G10/G11/G12) each
take a `model_available` flag that turns their "the model really streamed text /
really called a tool" assertions on or off. Phase 4 could set that flag from the
presence of OPENROUTER_API_KEY; Phase 5 is deliberately key-free (the pinned build
ships the free `opencode/big-pickle` default), so the flag has to come from
somewhere else.

Asking `GET /provider` whether it has a `default` is NOT that somewhere else - that
was run #13's heuristic and it cannot mean anything. Upstream fills `default` from
the models.dev catalog: Provider.defaultModelIDs()
(packages/opencode/src/provider/provider.ts:1132) maps EVERY catalogued provider to
`sort(Object.values(item.models))[0].id`, so the key is present whether or not a
provider can actually serve a turn; and `connected`
(packages/opencode/src/server/routes/instance/httpapi/handlers/provider.ts:60) is
`id in Provider.list() || auth.json[id]`, which depends on SDK init behaviour this
gate must not reason about indirectly.

So probe the only contract that matters: send one tiny prompt through OpenCode's own
`POST /session/:id/prompt_async` and look for an assistant message with text, using
nothing but the HTTP API a client would use (no privileged access, no model name
hardcoded - the server picks its own default).

Stdlib only (urllib), because it runs on the CI host before/independently of any
runner download, and talks to the server through the existing `adb forward`.

Exit 0 -> PROBE ok  : a real assistant reply arrived (model-dependent assertions on)
Exit 1 -> PROBE unavailable : error/timeout (those assertions stay off, loudly)
"""
import json
import os
import sys
import time
import urllib.error
import urllib.parse
import urllib.request
import base64

BASE = os.environ.get("OPENCODE_BASE", "http://127.0.0.1:4111")
USER = os.environ.get("OPENCODE_SERVER_USERNAME", "opencode")
PASS = os.environ.get("OPENCODE_SERVER_PASSWORD", "")
DIRECTORY = os.environ.get("OPENCODE_DIRECTORY", "")
TIMEOUT = int(os.environ.get("P5_MODEL_PROBE_TIMEOUT", "150"))
POLL = 3
TEXT = "Reply with exactly this token and nothing else: P5PROBEOK"


def call(method, path, body=None):
    url = BASE + path
    if DIRECTORY and path not in ("/global/health", "/global/config", "/global/event"):
        sep = "&" if "?" in path else "?"
        url += sep + "directory=" + urllib.parse.quote(DIRECTORY, safe="")
    data = None
    headers = {"Authorization": "Basic " + base64.b64encode(f"{USER}:{PASS}".encode()).decode()}
    if body is not None:
        data = json.dumps(body).encode()
        headers["content-type"] = "application/json"
    elif method != "GET":
        headers["content-length"] = "0"
    req = urllib.request.Request(url, data=data, headers=headers, method=method)
    try:
        with urllib.request.urlopen(req, timeout=30) as r:
            raw = r.read().decode("utf-8", "replace")
            return r.status, (json.loads(raw) if raw.strip() else None)
    except urllib.error.HTTPError as e:
        raw = e.read().decode("utf-8", "replace")
        try:
            return e.code, json.loads(raw)
        except Exception:
            return e.code, raw[:300]


def main():
    if not PASS:
        print("PROBE unavailable :: no server password (P5-04 did not export one)")
        return 1
    code, health = call("GET", "/global/health")
    if not (isinstance(health, dict) and health.get("healthy")):
        print(f"PROBE unavailable :: server not healthy (http {code} {str(health)[:120]})")
        return 1
    code, sess = call("POST", "/session", {"title": "P5 model pre-flight"})
    sid = sess.get("id") if isinstance(sess, dict) else None
    if not sid:
        print(f"PROBE unavailable :: POST /session -> http {code} {str(sess)[:200]}")
        return 1
    code, _ = call("POST", f"/session/{urllib.parse.quote(sid)}/prompt_async",
                   {"parts": [{"type": "text", "text": TEXT}]})
    if code not in (200, 204):
        print(f"PROBE unavailable :: prompt_async -> http {code} {str(_)[:200]}")
        return 1
    deadline = time.time() + TIMEOUT
    while time.time() < deadline:
        time.sleep(POLL)
        code, msgs = call("GET", f"/session/{urllib.parse.quote(sid)}/message")
        items = msgs if isinstance(msgs, list) else (msgs or {}).get("data") if isinstance(msgs, dict) else None
        if not isinstance(items, list):
            continue
        for m in items:
            info = m.get("info", m) if isinstance(m, dict) else {}
            parts = m.get("parts", []) if isinstance(m, dict) else []
            if info.get("role") != "assistant":
                continue
            err = info.get("error")
            if err:
                txt = json.dumps(err)[:300]
                print(f"PROBE unavailable :: assistant turn errored: {txt}")
                return 1
            for p in parts:
                if isinstance(p, dict) and p.get("type") == "text" and TEXT.split()[-1] in (p.get("text") or ""):
                    model = info.get("modelID") or info.get("model") or "?"
                    print(f"PROBE ok :: model={model} replied after <= {TIMEOUT - int(deadline - time.time())}s")
                    return 0
        # Text may have arrived without the exact token (a chatty model still proves
        # the turn ran); treat any non-empty assistant text as usable, but say so.
        for m in items:
            info = m.get("info", m) if isinstance(m, dict) else {}
            if info.get("role") != "assistant":
                continue
            for p in (m.get("parts") or []):
                if isinstance(p, dict) and p.get("type") == "text" and (p.get("text") or "").strip():
                    print("PROBE ok :: assistant text present (token not verbatim) "
                          f"model={info.get('modelID') or info.get('model') or '?'}")
                    return 0
    print(f"PROBE unavailable :: no assistant reply within {TIMEOUT}s")
    return 1


if __name__ == "__main__":
    sys.exit(main())

File: phase5/scripts/device/p5-model-probe.js
// p5-model-probe.js - decide, by measurement, whether a live agent turn can run.
//
// Why this exists: the Phase 5 re-runs of the Phase 3 drivers (G7/G10/G11/G12) and the
// in-app K2/K5 turns need to know whether a model can actually serve a prompt. Phase 4
// could infer that from the presence of OPENROUTER_API_KEY; Phase 5 is deliberately
// key-free (the pinned build ships the free `opencode/big-pickle` default), so the answer
// has to come from the server itself.
//
// Asking `GET /provider` whether it has a `default` cannot be that answer: upstream
// derives `default` from the models.dev catalog for EVERY provider
// (Provider.defaultModelIDs, packages/opencode/src/provider/provider.ts:1132 ->
// sort(Object.values(item.models))[0].id), so the key is present whether or not anything
// can serve a turn; and `connected`
// (packages/opencode/src/server/routes/instance/httpapi/handlers/provider.ts:60) is
// `id in Provider.list() || auth.json[id]`, which depends on SDK-init behaviour this gate
// must not reason about indirectly.
//
// So: send one tiny prompt through OpenCode's own public API and look for an assistant
// text part. No model name is hardcoded - the server picks its default.
//
// It runs ON DEVICE under the payload's own bun, as the app uid, because the OpenCode
// server is a child of the app process: from inside that process loopback is guaranteed
// reachable, while a host-side `adb forward` may not be (run #14: every host-side request
// died with ECONNRESET the moment an `am instrument` call replaced the app process).
//
// Prints exactly one verdict line:
//   PROBE ok :: model=<id>            -> model-dependent assertions may run
//   PROBE unavailable :: <reason>     -> they stay skipped, with the reason on record
const BASE = process.env.OPENCODE_BASE || "http://127.0.0.1:4111"
const USER = process.env.OPENCODE_SERVER_USERNAME || "opencode"
const PASS = process.env.OPENCODE_SERVER_PASSWORD || ""
const DIR = process.env.OPENCODE_DIRECTORY || ""
const DEADLINE = Date.now() + Number(process.env.P5_MODEL_PROBE_TIMEOUT || 150) * 1000
const START = Date.now()
const TOKEN = "P5PROBEOK"
const PROMPT = `Reply with exactly this token and nothing else: ${TOKEN}`

const auth = "Basic " + btoa(`${USER}:${PASS}`)

async function call(method, path, body) {
  const url = BASE + path + (path.includes("?") ? "&" : "?") + "directory=" + encodeURIComponent(DIR)
  const init = { method, headers: { authorization: auth } }
  if (body !== undefined) {
    init.headers["content-type"] = "application/json"
    init.body = JSON.stringify(body)
  }
  const r = await fetch(url, init)
  const text = await r.text()
  let json = null
  try { json = text ? JSON.parse(text) : null } catch { /* keep raw below */ }
  return { status: r.status, json, text: text.slice(0, 400) }
}

function bail(reason) {
  console.log(`PROBE unavailable :: ${reason}`)
  process.exit(1)
}

async function main() {
  if (!PASS) bail("no server password in the environment")
  const h = await call("GET", "/global/health")
  if (h.status !== 200 || h.json?.healthy !== true) bail(`health -> http ${h.status} ${h.text}`)
  const created = await call("POST", "/session", { title: "P5 model pre-flight" })
  const sid = created.json?.id
  if (!sid) bail(`POST /session -> http ${created.status} ${created.text}`)
  const sent = await call("POST", `/session/${encodeURIComponent(sid)}/prompt_async`, {
    parts: [{ type: "text", text: PROMPT }],
  })
  if (sent.status !== 200 && sent.status !== 204) bail(`prompt_async -> http ${sent.status} ${sent.text}`)

  while (Date.now() < DEADLINE) {
    await new Promise((r) => setTimeout(r, 3000))
    const got = await call("GET", `/session/${encodeURIComponent(sid)}/message`)
    const items = Array.isArray(got.json) ? got.json : (got.json?.data ?? [])
    for (const m of items) {
      const info = m.info ?? m
      if (info?.role !== "assistant") continue
      if (info.error) bail(`assistant turn errored: ${JSON.stringify(info.error).slice(0, 300)}`)
      const text = (m.parts ?? [])
        .filter((p) => p?.type === "text")
        .map((p) => p.text || "")
        .join("")
      if (text.includes(TOKEN)) {
        console.log(`PROBE ok :: model=${info.modelID || info.model || "?"} exact-token reply, session=${sid}`)
        return
      }
      if (text.trim()) {
        // A chatty model that did not echo the token verbatim still proves the turn ran.
        console.log(`PROBE ok :: model=${info.modelID || info.model || "?"} assistant text present (token not verbatim), session=${sid}`)
        return
      }
    }
  }
  bail(`no assistant reply within ${Math.round((Date.now() - START) / 1000)}s; session=${sid}`)
}

main().catch((e) => bail(`${e?.name || "error"}: ${e?.message || String(e)}`))
